REF: refactor sjoin as per #1463

adriangb opened the pull request:
This applies the structure discussed in #1463. This should make it easier to further develop `sjoin` in the future. Please feel free to suggest other names, docstrings or arrangements, mine are somewhat arbitrary.

martinfleis commented on geopandas/tools/sjoin.py:
```diff
@@ -101,10 +129,32 @@ def sjoin(
+    return result
+
+
+def _frame_join(result, left_df, right_df, how, lsuffix, rsuffix):
```
Can you include`lsuffix` and `rsuffix`  in the docstring?

martinfleis replied:
Same as above, can we not call this `results`?

martinfleis commented on geopandas/tools/sjoin.py:
```diff
@@ -28,7 +28,20 @@ def sjoin(
+    """
+    _basic_checks(left_df, right_df, how, op, lsuffix, rsuffix)
+
+    result = _geom_predicate_query(left_df, right_df, op)  # consume op param
```
Can we not call this intermediate df `results`? It can become confusing why there are other steps after results.

adriangb replied:
I renamed to `indices`

martinfleis commented on geopandas/tools/sjoin.py:
```diff
@@ -101,10 +129,32 @@ def sjoin(
+
+    Parameters
+    ----------
+    result : DataFrame returned by the geometric join.
```
And here as well.

adriangb commented:
Hi folks, thank you for the review! A lot of the issues are just carryover (even though the diff makes it look like I introduced it), but I am happy to fix anything here that is reasonably small. I'll start addressing the comments above, but feel free to leave more rename/docstring requests.

adriangb commented on geopandas/tools/sjoin.py:
```diff
@@ -101,10 +129,32 @@ def sjoin(
     # index in geopandas may be any arbitrary dtype. so reset both indices now
     # and store references to the original indices, to be reaffixed later.
     # GH 352
+    index_left = "index_%s" % lsuffix
```
This was a carryover, but I replaced all `%` with `format`. Hopefully someday we can just use f-strings.

martinfleis replied:
> Hopefully someday we can just use f-strings.

Right after we release 0.8.0 we drop support for Python 3.5. #1457

adriangb commented:
All comments should be addressed.

brendan-ward commented on geopandas/tools/sjoin.py:
```diff
@@ -6,6 +6,9 @@
 from geopandas.array import _check_crs, _crs_mismatch_warn
 
 
+ALLOWED_OP = {"intersects", "contains", "within"}
```
@adriangb it seems incongruous that `ALLOWED_OP` is here, whereas `allowed_hows` is within `_basic_checks` below.  (seems like things that perform a similar role should be in a similar place)

Even if ops are added in later refactors (e.g., based on new predicates from `query_bulk`), it seems like this could remain inline within `_basic_checks` unless you foresee needing to check `ALLOWED_OPS` in other places (in which case maybe that is the time to promote to a global variable?)

adriangb replied:
My hope is that in the future we offload this check/task to `sindex.query_bulk` since that is what really determines what `op` values are valid.* I felt that making it a global variable like this could help with that goal, but I do see your point. I'll move it back if nothing else to minimize the diff.

* I know this is off topic, but to just to start thinking about: aside from `within/contains` there is no `op/predicate` logic in `sjoin` anymore, so do we need tests for all of the other predicates before adding them? Or are we okay just having tests for `within`, `contains` and one of the other predicates (maybe `intersects` because tests are already implemented)? The rest of the testing would be up to `sindex` or, further down the pipeline, `pygeos`/`shapely2`.

adriangb replied:
Following up on this: perhaps the best option here is to not check the `op` parameter and let it be caught by `sindex` and bubble up. In the future, when `sindex` is user-defined and/or on a per-column basis, or also if we give users the ability to plug their own `_geom_predicate_query`, we would have to do it that way anyhow.

martinfleis replied:
>  perhaps the best option here is to not check the op parameter and let it be caught by sindex and bubble up

That sounds reasonable. Do we get a meaningful/compatible error message there?

brendan-ward replied:
In the case of pygeos, an invalid predicate (`foo`) raises:

```
ValueError: Predicate foo is not valid; must be one of within, crosses, contains, touches, overlaps, intersects
```
(other predicates will be added to list here as they are supported in pygeos)

brendan-ward replied:
> perhaps the best option here is to not check the op parameter and let it be caught by sindex and bubble up

This seems reasonable to me too.

> do we need tests for all of the other predicates before adding them?

I'm not sure about the `rtree` side, but we're striving for good unit test coverage of the predicates in pygeos - but we're still a bit weak on this across all geometry types.

The more I thought about this, the more I wanted the `sindex` tests in geopandas to be a bit more like integration tests.  Or put a bit differently, the unit tests in pygeos are using contrived geometries that have more easily known geometric relationships.  In geopandas, it might be good to verify at least some of these with geometries derived from real-world datasets, which would require a bit more manual review of the test outputs.

All that said, I think it is reasonable to rely on the core libraries to have good unit tests ("trust, but verify"), and add more thorough `sindex` tests in later PRs.

adriangb replied:
Totally agree @brendan-ward!

Regarding the error raised, we are currently checking the predicates in `sjoin` and a `ValueError` is raised for invalid predicates like so:
```
Got `predicate` = `foo`, `predicate` must be one of {None, within, crosses, contains, touches, overlaps, intersects}"
```
I think that is a reasonable error to see from `sjoin`. It may be a duplication of what `pygeos` is doing, but if we were to support more spatial index implementations we would want a unified interface, so I think having this duplication makes sense.

I'll implement this change tonight and update back here. Thanks all!

brendan-ward replied:
> having this duplication makes sense

agreed - thanks for following up on that, I didn't have a chance to check the latest in geopandas.

adriangb replied:
I realized that by removing this check, we would technically allow users to use other predicates, ex: `crosses`. I _expect_ it to work fine since it's just passing through to `sindex`, but there are no tests for this behavior. Do we need `sjoin` specific tests for this?

brendan-ward replied:
I'm not sure I follow.  If the predicates are validated at the `sindex` level, and they don't change the nature in which `sjoin` operates, then I don't think we need additional tests at the `sjoin` level to validate that those predicates are in the allowed list.

adriangb replied:
Exactly. That's my thought. I just want to make sure everyone is okay with that. 

To clarify: currently we have tests for `within`, `contains` and `intersects` in `sjoin`. `sjoin` _does_ have some logic regarding the `op` parameter, but since 3868b44e1a087565ba16ff228966f757bb58a7de it is limited to flipping `within/contains` or doing nothing otherwise. So I think it makes sense to keep the existing tests, but I do not think we need to add new ones. If we all agree on this (it sounds like you do @brendan-ward) we should go ahead and enable the rest of the predicates by dropping the check in `sjoin` and relying on `sindex` to decide what it can and cannot do.

The reason I think this refactor is the place to make this change is to avoid introducing the new `_basic_checks` function with `op` in it's signature only to change that a couple of commits later.

adriangb replied:
Okay seems like there is consensus! @jorisvandenbossche please chime in real quick if you can.

What I'll do is:
1. Make a new PR to adjust the `sjoin` documentation to reflect this change and assign it to myself.
2. Push the change to this PR to implement.

brendan-ward commented on geopandas/tools/sjoin.py:
```diff
@@ -28,7 +31,18 @@ def sjoin(
+
+
+def _basic_checks(left_df, right_df, how, op, lsuffix, rsuffix):
+    """Checks validity of input parameters.
```
Please consider expanding the docstring (feel free to edit):

```
Checks the validity of join input parameters.  
`how` and `op` must be one of the valid options.  
`'index_'` concatenated with `lsuffix` or `rsuffix` must not already exist as columns in the left or right data frames.

Parameters
------------
left_df : GeoDataFrame
right_df : GeoData Frame
how : str, one of 'left', 'right', 'inner'
    join type
op : str, one of 'intersects', 'contains', 'within'
    predicate operation used in join
lsuffix : str
   left index suffix
rsuffix : str
   right index suffix
```

adriangb replied:
Will do.

adriangb replied:
Expanded docstring, resolving.

martinfleis commented on geopandas/tools/sjoin.py:
```diff
@@ -43,20 +71,20 @@ def sjoin(
         )
 
     if not _check_crs(left_df, right_df):
         _crs_mismatch_warn(left_df, right_df, stacklevel=3)
```
```suggestion
        _crs_mismatch_warn(left_df, right_df, stacklevel=4)
```
I guess this needs to be 4 now. Can you check that?

adriangb replied:
Yes, it should be 4. I changed it to 4 now.

adriangb commented on geopandas/tools/sjoin.py:
```diff
@@ -28,7 +28,35 @@ def sjoin(
+    return joined
+
 
+def _basic_checks(left_df, right_df, how, op, lsuffix, rsuffix):
```
Reminder to remove this parameter and check.

adriangb replied:
Done.

brendan-ward commented on geopandas/tools/sjoin.py:
```diff
@@ -42,7 +39,24 @@ def sjoin(
+    op : str, one of 'intersects', 'contains', 'within'
+        predicate operation used in join
+    lsuffix : str
+    left index suffix
```
Tiny fix: please indent 57 & 59

adriangb replied:
Will do, thanks!

adriangb replied:
Fixed. Thanks again!

adriangb commented:
> Other than indents, this looks good. Thanks @adriangb !

Thanks!

Are we all okay with how I worded the documentation? I tried to strike a balance between saying "valid `op` values depend on the spatial index, so go check your spatial index" and "these are the valid values and they will never change". In any case, I think this is temporary because it will all have to be changed if/when we have separate sindex documentation (#1479) and/or make more moves on #1344 or #1464.

martinfleis commented:
> Are we all okay with how I worded the documentation?

That is fine. We'll certainly touch that part of docs in the following months anyway.

adriangb commented:
@martinfleis please take a look, I think this should be ready to merge. Thanks!

martinfleis commented:
Thanks @adriangb and @brendan-ward!
